File: app/src/main/java/com/veyndan/paper/reddit/post/media/mutator/TextMutatorFactory.java
package com.veyndan.paper.reddit.post.media.mutator;

import android.text.TextUtils;

import com.veyndan.paper.reddit.api.reddit.model.PostHint;
import com.veyndan.paper.reddit.post.media.model.Text;
import com.veyndan.paper.reddit.post.model.Post;

import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;

final class TextMutatorFactory implements MutatorFactory {

    static TextMutatorFactory create() {
        return new TextMutatorFactory();
    }

    private TextMutatorFactory() {
    }

    @Override
    public Maybe<Post> mutate(final Post post) {
        return Single.just(post)
                .filter(post1 -> post1.postHint().value == PostHint.SELF &&
                        !TextUtils.isEmpty(post1.bodyHtml()))
                .map(post1 -> {
                    final Text text = Text.create(post1::getDisplayBody);
                    return post1.withMedias(post1.medias().value.concatWith(Observable.just(text)));
                });
    }
}

File: app/src/main/java/com/veyndan/paper/reddit/post/media/mutator/TextMutatorFactory.kt
package com.veyndan.paper.reddit.post.media.mutator

import com.veyndan.paper.reddit.api.reddit.model.PostHint
import com.veyndan.paper.reddit.post.media.model.Text
import com.veyndan.paper.reddit.post.model.Post
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single

class TextMutatorFactory : MutatorFactory {

    override fun mutate(post: Post): Maybe<Post> {
        return Single.just(post)
                .filter { it.postHint == PostHint.SELF && !it.bodyHtml.isNullOrEmpty() }
                .map { post1 ->
                    val text: Text = Text { context -> post1.getDisplayBody(context)!! }
                    post1.copy(post1.medias.concatWith(Observable.just(text)))
                }
    }
}
